Advanced audit logging API

timstclair opened the pull request:
This is an initial implementation of the API proposed in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auditing.md

This deviates from the exact API specified in the proposal in a number of ways. Here's a list of some of the changes I made, and why:

- General API conformance (`metav1.{Type,List}Meta`, +optional, etc.)
- Dropped fields that won't be part of the initial implementation (storage objects & filters), to be added later
- {Request,Response}Body -> {Request,Response}Object - This is actually what the proposal originally had, and will be much better for adding Filters later, and provides a less protocol-dependent API. Note that there is currently no way to include Response body for non-resource requests.
- Drop HttpMethod - redundant with verb
- Response{StatusCode,ErrorMessage} collapsed into `ResponseStatus *metav1.Status` - This is the standard response type for resourec errors, but can hold genereric error responses too. Note that this is part of the `Metadata` Level.

/cc @kubernetes/sig-auth-api-reviews @kubernetes/api-reviewers

tallclair committed:
Add internal audit API types

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45315)
<!-- Reviewable:end -->

deads2k commented:
I think that auditing should be generic functionality, so I think these types deserve to go into `k8s.io/apiserver`.

sttts commented on pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,166 @@
+	LevelRequestObject Level = "RequestObject"
+	// LevelResponseObject provides RequestObject level of auditing, and additionally
+	// logs the response object (does not apply for non-resource requests).
+	LevelResponseObject Level = "ResponseObject"
```
StorageObject left out intentionally? Also in the proposal we talked about body (i.e. serialized data, not runtime.Objects).

timstclair replied:
See the PR description for an explanation. To elaborate a bit more on serialized data vs. runtime Objects:

My concern with serialized data is that it's protocol dependent. For instance, I'm assuming we don't want to log binary format requests, so if we get a protobuffer request, we would need to deserialize it, and then reserialize into json. Why not just include the deserialized version?

soltysh commented on pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,166 @@
+	Timestamp metav1.Time
+	// Unique audit ID, generated for each request.
+	// +optional
+	RequestID string
```
Either `AuditID`, or just `ID`, I'm worried `RequestID` might be confusing.

timstclair replied:
Done.

sttts commented on pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,166 @@
+	// RequestObject logged before admission. Omitted for non-resource requests.
+	// Only logged at RequestObject Level and higher.
+	// +optional
+	RequestObject runtime.Object
```
What about PATCH requests?

sttts replied:
Please add a comment that these objects will be external types before (request) and after (response) conversion.

timstclair replied:
If this is the external versioned object, should it be a `runtime.RawExtension` rather than a `runtime.Object`?

sttts replied:
It's not serialized, it's only converted from internal to external. This means it's not json yet. I.e. the audit backend has to serialize them into whatever format it prefers.

In the proposal we have runtime.Unstructured. This would mean that the backend gets parsed essentially json. Though this needs two conversions to create (externalal -> json blob -> Unstructured).

Having runtime.Object here is cheap as we have to create that anyway. The audit backend can whatever it likes with it.

sttts replied:
In short: I would keep `runtime.Object` for now.

soltysh commented:
> Reverted GroupResource change in policy - See kubernetes/community#145 (comment):

Commented in the original.

sttts commented on pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,166 @@
+	// Only logged at RequestObject Level and higher.
+	// +optional
+	RequestObject runtime.Object
+	// ResponseObject logged before admission. Omitted for non-resource requests.
```
"logged before admission" doesn't make sense here

timstclair replied:
Oops, copy pasta error. Fixed.

sttts commented on pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,166 @@
+	// A ResourceKind can be matched in any of the APIGroups listed above.
+	// Any empty list implies every resource kind.
+	// +optional
+	ResourceKinds []string
```
Agreeing with @deads2k: please use `schema.GroupResources` here instead of APIGroups+ResourceKinds.

timstclair replied:
Will do, but I'm going to go with the suggestion of having multiple kinds per group, e.g.:
```go
type GroupKinds struct {
    Group string
    Kinds []string
```

QUESTION: Should `ResourceNames` be moved into this object? I feel like `ResourceNames` faces the same problems as kinds to groups. The "correct" approach seems like it would be to replace `Kinds` with a `Resource struct{ Kind string, ResourceNames []string }`, but I'm concerned that's too heavy weight for now. Can we just remove `ResourceNames` for now?

soltysh replied:
👍 for removing `ResourceNames` for now.

timstclair commented:
> I think that auditing should be generic functionality, so I think these types deserve to go into `k8s.io/apiserver`.

I discussed offline with @lavalamp and we thought it made sense for audit logging to be the responsibility of the aggregator rather than user apiservers. However I agree that it makes sense for generic apiservers to have the capability of audit logging, so I will make the move.

timstclair commented:
Thanks for the feedback! I think I addressed everything. Open question about what to do with `ResourceNames`

ericchiang commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,173 @@
+	// *s are allowed, but only as the full, final step in the path.
+	// Examples:
+	//  "/metrics" - Log requests for apiserver metrics
+	//  "/healthz*" - Log all health checks
```
nit: I think RBAC requires you to include a slash before the wildcard for non-resource URLs, so for RBAC this would be:

```
/healthz
/healthz/*
```

Can we update to be consistent?

timstclair replied:
AFAICT that's not the case - there isn't any special [validation](https://github.com/kubernetes/kubernetes/blob/97295740e398b053da12565fdfa8a13c54053f8a/pkg/apis/rbac/validation/validation.go#L83), and the matching is just done with a [string prefix check](https://github.com/kubernetes/kubernetes/blob/97295740e398b053da12565fdfa8a13c54053f8a/pkg/apis/rbac/helpers.go#L93).

liggitt replied:
RBAC does not require a slash, so if the intent is to allow `/foo` and subpaths, two distinct patterns are required. `/foo*` would also allow `/foobar`

ericchiang reviewed (commented):
thanks for putting this together. just a nit but otherwise works for me.

soltysh commented:
I'm fine with leaving `ResourceNames` behind, for now at least and proceed with the current state.

timstclair commented:
Now that we've had a few back-and-forths, I'll work on getting the rest of this PR (boilerplate) together. I'm out tomorrow (5/5), but will try to have something by EOD Monday.

timstclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,169 @@
+	// +optional
+	ResourceKinds []GroupKinds
+	// Namespaces that this rule matches.
+	// The empty string "" matches non-namespaced resources.
```
Question to reviewers: does this make sense (`The empty string "" matches non-namespaced resources.`)? Or should we say that the `Namespaces` field is ignored for non-namespaced resources? The argument for the latter is that if I include non-namespaced resources in the `ResourceKinds` I probably meant to catch them, even if I forgot to add the `""` namespace. OTOH, I might want a way to say "log all resources in the namespace foo" without needing to enumerate every namespaced resource.

Actually, now that I wrote it out, I think the second point is a strong argument for keeping it as is (`""` for non-namespaced).

soltysh replied:
👍 for `""` for non-namespaced.

ericchiang replied:
works for me

timstclair commented:
Updated with the versioned code, boilerplate, and generated files. This PR is now ready for full review (PTAL).

timstclair commented:
hmm, still a few kinks to work out of the generated code... will pick it up again tomorrow.

timstclair commented:
@deads2k 

> I think that auditing should be generic functionality, so I think these types deserve to go into `k8s.io/apiserver`.

I should have caught this earlier, but `k8s.io/apiserver` isn't allowed to depend on `k8s.io/kubernetes`. Since the UserInfo type is specific to Kubernetes, we can't include it here. This leaves a couple options:

1. Redefine `UserInfo` in the audit API
2. Move the audit API back into `k8s.io/kubernetes`
3. Use a generic `RawExtension` type to let implementations fill in custom authn info

`ObjectReference` is also forbidden, but I feel less bad about cloning this, or moving it to the metav1 package. Thoughts?

timstclair commented:
Actually, I don't think (3) is compatible with the policy proposal, since that needs access to the username & groups.

timstclair commented:
OK, I just chatted with @caesarxuchao and this is a lot more complicated than I initially realized, since everything is being shuffled around. Thinking more broadly about where we want these APIs to live, there are a couple options:

1. Move EVERYTHING to `k8s.io/api`, including the audit API. If this is the long term goal, we should just implement it in `k8s.io/kubernetes` for now.
2. Should authentication be part of the `apiserver` API? If so, we can keep audit in `apiserver`, and just move the authentication.UserInfo in.
3. Alternatively, once the api is copied out to `k8s.io/api` there will no longer be the `k8s.io/kubernetes` --> `k8s.io/apiserver` --> `k8s.io/kubernetes` circular dependency, so audit could stay where it is and depend on `k8s.io/api/pkg/apis/authentication/v1.UserInfo`. If we go this route, we could temporarily depend on the copy in `client-go` until the move is complete. This option doesn't seem right to me...

I will follow up with @lavalamp and @caesarxuchao tomorrow, but if you have thoughts on this please chime in here.

liggitt commented:
cc @deads2k and @sttts on dependency cycle question

sttts commented:
Please don't move this to `k8s.io/api`. It has nothing to do with the actual API types, it's an apiserver topic. No client should care about auditing.

I vote for 3. We do this in couple of places already to depend on the client-go variant of types. And as these will merge with k8s.io/api soon this is a nobrainer IMO, if this breaks the import cycle for now.

sttts commented:
> This option doesn't seem right to me...

@timstclair why?

soltysh commented:
👍 for *not* adding this to `k8s.io/api` and option no 3. @sttts confirmed that we're already doing this in several places so I'm fully supportive for that direction.

caesarxuchao commented:
If we go for no. 3, note that the internal audit types can't depend on the types in client-go/pkg/api, instead, it should duplicate the type definition, like what we did in #45384, because client-go/pkg/api will be deprecated soon. For the external audit types, it's fine to depend on client-go/pkg/api/v1, which will be moved to k8s.io/api in the future.

sttts commented:
I am also fine with duplicating `UserInfo` in k8s.io/apiserver. The type is not that complex:

```golang
type UserInfo struct {
	Username string `json:"username,omitempty" protobuf:"bytes,1,opt,name=username"`
	UID string `json:"uid,omitempty" protobuf:"bytes,2,opt,name=uid"`
	Groups []string `json:"groups,omitempty" protobuf:"bytes,3,rep,name=groups"`
	Extra map[string]ExtraValue `json:"extra,omitempty" protobuf:"bytes,4,rep,name=extra"`
}

type ExtraValue []string
```
Also, there is no dependency on pkg/api.

I feel pretty strong about keeping everything in k8s.io/apiserver.

In general about authentication we decided to assume that our apiserver library lives in a kube environment, i.e. as a sidekick apiserver to kube's apiserver. So having a dependency on client-go is fine in general. But, I agree that it is a bit strange to reference a client type in a public apiserver type. Also because we might have no control over the generated code of an imported type in the furtue (now we still have as we live in staging in the same repo. But some day we won't anymore).

timstclair commented:
To clarify my early comments:

> Move EVERYTHING to k8s.io/api, including the audit API. If this is the long term goal, we should just implement it in k8s.io/kubernetes for now.

By EVERYTHING I just meant the api types (everything in `pkg/apis/audit`). I'm not suggesting moving the audit machinery.

> If we go this route, we could temporarily depend on the copy in client-go until the move is complete. This option doesn't seem right to me...

I meant that I think it's odd for the apiserver to depend on a client package.

timstclair commented:
@caesarxuchao For some reason changing the dependencies to k8s.io/client-go from k8s.io/kubernetes causes hack/update-generated-protobuf.sh to create a UserInfo message, but the generated.pb.go is broken since there is no UserInfo struct.... Any ideas what's going on?

/cc @smarterclayton

timstclair commented:
@soltysh @sttts @deads2k I really want to unblock the integration work. If I remove the proto generation for now, are there any other issues with this PR or can we move forward?

sttts commented:
@deads2k approved?

sttts commented:
@smarterclayton any idea what goes wrong here that protobuf-gen leaves in the UserInfo message? Tried to find out, but got lost in our protobuf post-processing steps.

sttts commented:
@timstclair only question left: why do you redefine UserInfo in the internal types, but use the client-go variant in v1alpha1? Other than that, I am fine with moving forward.

soltysh commented:
👍 on my end after answering @sttts question

timstclair commented:
Because @caesarxuchao told me to :smile: 

> If we go for no. 3, note that the internal audit types can't depend on the types in client-go/pkg/api, instead, it should duplicate the type definition, like what we did in #45384, because client-go/pkg/api will be deprecated soon. For the external audit types, it's fine to depend on client-go/pkg/api/v1, which will be moved to k8s.io/api in the future.

My understanding is that in the future internal types aren't intended to be "exported", and shouldn't cross repo boundaries.

sttts commented:
> My understanding is that in the future internal types aren't intended to be "exported", and shouldn't cross repo boundaries.

Ok, understood now. Please make sure that the conversions are correct.

I have a small tendency to prefer a duplicated type also in the external types. But I let you decide. Lgtm so far (when the protobuf issues is fixed or we skip protobuf for now). Hint: with a duplicated type in the external types protobuf-gen will most probably be happy.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,201 @@
+type Event struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
```
sorry for the late comment: why do we need ListMeta? This is not going to be served via an api group endpoint. Neither SelfLink nor ResourceVersion make much sense.

sttts replied:
If we want to make it into a api group of an external audit apiserver, it should be ObjectMeta, same for Policy below.

timstclair commented:
Updates:

1. I think we actually don't want to generate a client for the audit types. The only operation that needs to be done on the types is posting to the webhook (a create). Looking at how this is done for the authorizer, that just ends up creating a custom Create client anyway: https://github.com/kubernetes/apiserver/blob/53a5f2e3f3cce9d58f42200c2bbee1980ec8e96b/plugin/pkg/authorizer/webhook/webhook.go#L210-L242 (thanks @ericchiang )
2. Changed the Request & ResponseObject types to a JSON `string`. This is because we want json format even if the audit event is passed in proto format, since the endpoint may not understand the object types (proto is not self-describing). I also renamed the audit level & field names to be consistent with this change.

I've also forked off https://github.com/kubernetes/kubernetes/pull/45862 to try and unblock the implementation with internal types while we sort out the code generation.

timstclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,200 @@
+)
+
+// Level defines the amount of information logged during auditing
+type Level string
```
Yes, but we don't do that anywhere else in our APIs (search `file:pkg/api iota`). Is it ok if we stick with the convention for now, and reconsider it for the internal type later?

timstclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,200 @@
+	LevelMetadata Level = "Metadata"
+	// LevelRequestObject provides Metadata level of auditing, and additionally
+	// logs the request object (does not apply for non-resource requests).
+	LevelRequestObject Level = "RequestObject"
```
Yeah... we discussed this back in the original proposal [1]. The conclusion there was that this was probably an uncommon use case, and the request could be filtered out (once filters are implemented) if really desired. From an implementation perspective, I think this works since auditing the response is strictly more work on the system than auditing the request.

That said, we could rename the `Response` level to `RequestResponse` (to indicate that both are logged), then if there were sufficient demand we could retroactively add `Response` (_only_ the response) without any other api changes. WDYT?

[1] https://github.com/kubernetes/community/pull/145#discussion_r112614337

timstclair commented:
Ugh, github decided I didn't need to respond to https://github.com/kubernetes/kubernetes/pull/45315#discussion_r116631694 (no reply field), so here it is:

> Wondering if a something like a negated namespace would be nice. If I want an audit level of X for kube-system and Y for everything else it seems painful to maintain.

If you really want level X for kube-system and level Y for everything else that's actually pretty easy:

```yaml
- Level: X
  Namespaces: ["kube-system"]
- Level: Y
```

Or more generally:

```yaml
- Level: X
  Namespaces: ["kube-system"]
  <other_constraints>
- Level: Y
  <other_constraints>
```

Actually, I was trying to come up with a good counter example where this would be tricky, and it was hard enough that I convinced myself we don't need negated namespaces :) Again, with sufficient demand I think we can add it retroactively. WDYT?

timstclair commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -0,0 +1,200 @@
+// user.Info interface.
+type UserInfo struct {
+	// The name that uniquely identifies this user among all active users.
+	Username string
```
This will be filled based on the authorizer attributes, do you know what is used in that case? We do have rbac `ClusterRoleBindings` for controllers, e.g. 

```yaml
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:controller:daemon-set-controller
subjects:
- kind: ServiceAccount
  name: daemon-set-controller
  namespace: kube-system
```

I was thinking the Username in this case would be `daemon-set-controller`. Should we clarify that Username may refer to a service account?

If there actually isn't an authenticated user, we still use the `system:anonymous` user. The one case where this might actually be empty is for requests to the insecure local port (bleh). We should probably define a magic string to use in that case... `system:unauthenticated`?

/cc @liggitt

liggitt replied:
Requests should *always* have user info associated. Even the unsecured port now adds user.Info to the context as of https://github.com/kubernetes/kubernetes/pull/43888

liggitt replied:
Controllers are users too... they either use the credentials given to the controller-manager via --kubeconfig, or they use individual service account credentials, or they come in via the unsecured port and get the `system:unsecured` username assigned

liggitt replied:
The username for a service account is `system:serviceaccount:<sa-namespace>:<sa-name>`

timstclair commented on staging/src/k8s.io/apiserver/pkg/client/clientset_generated/clientset/clientset.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// NewForConfig creates a new Clientset for the given config.
+func NewForConfig(c *rest.Config) (*Clientset, error) {
```
(generated client removed)

timstclair commented on staging/src/k8s.io/apiserver/pkg/client/clientset_generated/clientset/fake/register.go:
```diff
@@ -0,0 +1,53 @@
+limitations under the License.
+*/
+
+package fake
```
(generated client removed)

timstclair commented on staging/src/k8s.io/apiserver/pkg/client/informers/externalversions/audit/v1alpha1/policy.go:
```diff
@@ -0,0 +1,68 @@
+package v1alpha1
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
(generated client removed)

I generally agree with this in non-generated code though, with the caveat that package names should [not have underscores](https://blog.golang.org/package-names) (should be `metav1`).

timstclair commented:
Clarified offline that PolicyRules are strictly ordered. Should this be clarified somewhere?

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,170 @@
+
+	// Resource kinds that this rule matches. An empty list implies all kinds in all API groups.
+	// +optional
+	ResourceKinds []GroupKinds `json:"resourceKinds,omitempty" protobuf:"bytes,5,rep,name=resourceKinds"`
```
We need s/Kind/Resource/ here. On the meta data level (e.g. in the aggregator) we don't know the kind, only the GroupResource.

timstclair replied:
There are a few changes to the API we need now... would it be ok to submit this (more or less) as is, and then follow up with the api changes?

deads2k replied:
> There are a few changes to the API we need now... would it be ok to submit this (more or less) as is, and then follow up with the api changes?

I'd stick on this.  It's definitely a resource, not a kind at the point where we make an auditing decision.

timstclair commented:
I'm going to rebase & squash to clean this up.

tallclair committed:
Update generated files

tallclair committed:
Misc proto changes

timstclair commented:
Ok, squashed & rebased. There are a couple of API changes we will need to make, but I'd like to submit this as is to capture all the boilerplate, and then follow up with the API changes.

PTAL

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/install/install.go:
```diff
@@ -0,0 +1,46 @@
+			GroupName:                  audit.GroupName,
+			VersionPreferenceOrder:     []string{v1alpha1.SchemeGroupVersion.Version},
+			ImportPrefix:               "k8s.io/apiserver/pkg/apis/audit",
+			RootScopedKinds:            sets.NewString(),
```
I would have expected audit types to be cluster scoped.  Are they not?

timstclair replied:
done

sttts replied:
Didn't we decide that we won't make that into a real CRUD api group (yet)? I still think we should drop install.go (and this decision about cluster scope) until we do.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/register.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+var (
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
```
copy/paste remnant?

timstclair replied:
yup, removed.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/register.go:
```diff
@@ -0,0 +1,59 @@
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&Event{},
+		&EventList{},
+		&Policy{},
```
I'm a little surprised there's no list, but I haven't delved deep yet.

timstclair replied:
Policy is top level configuration, so I don't think it makes sense for it to have a list. Put another way, we could think of the Policy as the list type, since it just holds a list of PolicyRules...

Does that answer your question?

soltysh replied:
For simplicity reasons (the generation tools) you might want to add one. The tools will generate `List/Watch` methods, because all the resources available in k8s always have a list. I've run into this problem some time ago working with generated code in openshift. We either need to fix generation or add list, here.

timstclair replied:
Added `PolicyList`

sttts replied:
The client needs it. No client => we don't need PolicyList.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+// Event captures all the information that can be included in an API audit log.
+type Event struct {
```
Lost of infrastructure expects `ObjectMeta`.  I don't see much value in fighting that infrastructure.  I'm good with including ObjectMeta like we did for SARs

timstclair replied:
done

sttts replied:
Looks like we go back and forth here. I am still thinking it's not necessary and even misleading for non-CRUD types. But if you insist, I won't object.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	// merging. It is an external versioned object type, and may not be a valid object on its own.
+	// Omitted for non-resource requests.  Only logged at RequestObject Level and higher.
+	// +optional
+	RequestBody string `json:"requestBody,omitempty" protobuf:"bytes,12,opt,name=requestBody"`
```
If this is json format (says it is in the doc), then I think this should be a `runtime.RawExtension`

timstclair replied:
Done. I'm still unsure whether the internal type (`runtime.Unkown`) is correct though.

sttts replied:
It shouldn't be runtime.RawExtension. How will RawExtension be encoded in the protobuf case? As protobuf? Then we are lost in the webhook because we cannot decode it and want probably store json anyway.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	// to the external type, and serialized as JSON.  Omitted for non-resource requests.  Only logged
+	// at ResponseObject Level and higher.
+	// +optional
+	ResponseBody string `json:"responseBody,omitempty" protobuf:"bytes,13,opt,name=responseBody"`
```
`runtime.RawExtension`

timstclair replied:
done

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	// RequestURI is the request URI as sent by the client to a server.
+	RequestURI string `json:"requestURI" protobuf:"bytes,5,opt,name=requestURI"`
+	// Verb is the kubernetes verb associated with the request.
+	// For non-resource requests, this is identical to HttpMethod.
```
I think you should be opinionated on case.  Pretty sure we `ToLower` these in the authorizer

timstclair replied:
done

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	// Time the request reached the apiserver.
+	Timestamp metav1.Time `json:"timestamp" protobuf:"bytes,3,opt,name=timestamp"`
+	// Unique audit ID, generated for each request.
+	AuditID types.UID `json:"auditID,omitempty" protobuf:"bytes,4,opt,name=auditID,casttype=k8s.io/apimachinery/pkg/types.UID"`
```
I'm convinced this should be an ordered slice.

soltysh replied:
Why ordered slice? Each event has it's own unique id. Even if we'll be talking about aggregation and federated servers, we'll have some sort of global ID coming from aggregator/federation and a unique one for each event. This is the latter, for now at least.

timstclair replied:
Agree with @soltysh. I think the Audit ID + slice of source IPs should provide the same information as a slice of audit IDs.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	// Rules specify the audit Level a request should be recorded at.
+	// A request may match multiple rules, in which case the FIRST matching rule is used.
+	// The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+	Rules []PolicyRule `json:"rules" protobuf:"bytes,2,rep,name=rules"`
```
This construct suggests to me that the API type is a `PolicyRule` and the level of config is done via a `PolicyRuleList`

soltysh replied:
Yeah, going this would definitely solve the problem of list generation I've mentioned before. +1

timstclair replied:
I'm worried that making Policy a list type makes the type to rigid (not future proof). Also I don't think PolicyRule conceptually makes sense as a standalone type, especially since they're strictly ordered in a policy. I added a `PolicyList` instead.

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+	User authnv1.UserInfo `json:"user" protobuf:"bytes,7,opt,name=user"`
+	// Impersonated user information.
+	// +optional
+	Impersonate *authnv1.UserInfo `json:"impersonate,omitempty" protobuf:"bytes,8,opt,name=impersonate"`
```
`impersonatingUser`?  `impersonatee`?

timstclair replied:
`ImpersonatedUser`?

ericchiang replied:
`impersonating`?

soltysh replied:
+1 for the `ImpersonatedUser`

timstclair replied:
done

deads2k commented on staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// GroupKinds represents resource kinds in an API group.
+type GroupKinds struct {
```
Needs to be GroupResources

timstclair replied:
done

sttts replied:
:+1:

timstclair commented:
Thanks for the review. Ready for another pass. PTAL.

soltysh commented:
LGTM, the only question is why you dropped protobuf tags?

timstclair commented:
> the only question is why you dropped protobuf tags?

Those tags were autogenerated by the protobuf generating tool, and we dropped protobuf generation (only from this PR, since I was running into problems). The tags will be added back once we fix protobuf generation.

soltysh commented:
👍 thanks!

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -75,14 +79,14 @@ type Event struct {
+	// Omitted for non-resource requests.  Only logged at Request Level and higher.
 	// +optional
-	RequestBody string
+	RequestObject runtime.Unknown
```
How does this help with filters? Filters will not be done by the backend, but before writing it into the event.

sttts replied:
I would suggest to use runtime.Object in the internal type and have a serialization to string during conversion to the external type. @deads2k wdyt?

deads2k replied:
> I would suggest to use runtime.Object in the internal type and have a serialization to string during conversion to the external type. @deads2k wdyt?

I think I'll leave the exact mechanism to be the easiest available, but I wanted to make certain it wasn't a string.  I think that `runtime.Unknown` will be guaranteed to pass straight through without anything trying to encode or decode directly.  If we use a `runtime.Object` it will be affected by the outer encoding structure (again, I think).  `runtime.RawExtension` may work internal and external as well.

I really just wanted to guarantee this didn't end up as a string so that we'd be able to directly inspect with a jsonpath filter or the like against incoming data if we wished without having to pull out the information.  Someday jsonpath may catch up to xml and we'll get streaming evaluators.

sttts replied:
The jsonpath idea is a good point. If `runtime.Unknown` comes through as a sub-expression of the object, that's perfect. This is also true for protobuf, i.e. if the `Event` is serialized as protobuf, will the `runtime.Unknown` come through correctly, i.e. either as JSON or as some protobuf encoding of JSON?

deads2k replied:
> How does this help with filters? Filters will not be done by the backend, but before writing it into the event.

I'd like the object shape to be inspectable if the sync does something like write it directly to a different etcd. Not nice, not good, but having the object directly embedded means that anyone looking at an `Event.audit.k8s.io` can run jsonpath and jq sanely against the object itself. I don't think that stringifying json inside of (likely) json is path we'd be happy with.  We encouraged servicecatalog to avoid it for much the same reason.

deads2k replied:
> This is also true for protobuf, i.e. if the Event is serialized as protobuf, will the runtime.Unknown come through correctly, i.e. either as JSON or as some protobuf encoding of JSON?

`runtime.Unknown` should come through as its native encoding.  So the event fields would be protobuf, but this element would be json inside of it.  We use this for `Template.templates.openshift.io` which have an outer proto encoding with an inner json one.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -98,13 +102,26 @@ type EventList struct {
 	metav1.TypeMeta
+	// ObjectMeta is included for interoperability with API infrastructure.
+	// +optional
+	metav1.ObjectMeta
```
why? Which infrastructure?

deads2k replied:
> why? Which infrastructure?

Generation, generic clients, scope naming, API installer infrastructure (if we ever want it).  I've been down the path of excluding this before and you get whacky errors.  I don't see a reason to fight our infrastructure on a few fields, several of which we probably want in the end like annotations.

sttts replied:
Ok, all the arguments I had with my previous pull https://github.com/kubernetes/kubernetes/pull/45765. I don't argue against this, quite the opposite!

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -33,14 +34,17 @@ const (
 	metav1.TypeMeta
+	// ObjectMeta is included for interoperability with API infrastructure.
+	// +optional
+	metav1.ObjectMeta
```
why?

deads2k replied:
> why? Which infrastructure?

Generation, generic clients, scope naming, API installer infrastructure (if we ever want it).  I've been down the path of excluding this before and you get whacky errors.  I don't see a reason to fight our infrastructure on a few fields, several of which we probably want in the end like annotations.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -33,14 +34,17 @@ const (
+	// LevelRequestResponse provides Request level of auditing, and additionally
 	// logs the response object (does not apply for non-resource requests).
-	LevelResponse Level = "Response"
+	LevelRequestResponse Level = "RequestResponse"
```
They are ordered anyway. To be consistent this must be LevelMetaRequestResponse. No I don't like that.

timstclair replied:
The idea was that you the user would always want metadata, and it's just the request/response part that doesn't have a well defined "ordering". Anyhow, let's follow up elsewhere.

deads2k commented:
These types lgtm. If the `runtime.Unknown` proves difficult to work with, but @smarterclayton or myself. If we need to switch to `runtime.RawExtension` we can.  I just wanted to make sure we had the object embedded, not stringified.

sttts commented:
With https://github.com/kubernetes/kubernetes/pull/45315#pullrequestreview-38857804 being a nit and `runtime.Unknown` to be what we want (if not, we can change that in a follow-up), let's merge this and iterate.

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -75,14 +79,14 @@ type Event struct {
+	// at Response Level.
 	// +optional
-	ResponseBody string
+	ResponseObject runtime.Unknown
```
Just noticed: runtime.Unknown is a struct, not an interface. So detection whether this field is set, is a bit awkward. Maybe we should make it a pointer (assuming the encoding stays the same in json).

timstclair replied:
Yeah, I originally had it as a pointer type (both the runtime.Unknown and runtime.RawExtension) but I noticed that everywhere else it's used (well, runtime.Unknown doesn't appear to be used directly anywhere) in our code without a pointer, even where it's optional (e.g. [here](https://github.com/kubernetes/kubernetes/blob/e9b02c2e2b206ab0f4beb146e5fd1edcc385b70b/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go#L47) and [here](https://github.com/kubernetes/kubernetes/blob/e9b02c2e2b206ab0f4beb146e5fd1edcc385b70b/pkg/watch/json/types.go#L41))

smarterclayton replied:
For internal types, runtime.Unknown will be "set" if internal data is nil or not.

timstclair commented:
Squashed & sync'd, ready for merge?

deads2k commented:
/lgtm

thockin commented:
/approve

timstclair commented:
@k8s-bot gce etcd3 e2e test this

timothysc commented:
/cc @kensimon

timstclair commented:
Feature: https://github.com/kubernetes/features/issues/22
